File: src/ghostdesk/instructions.py
# Copyright (c) 2026 Yoann Vanitou — FSL-1.1-ALv2
"""LLM instructions for the GhostDesk MCP server."""

INSTRUCTIONS = """
You control a virtual Linux desktop through screen capture, mouse, and
keyboard. Each tool's docstring covers its own parameters and caveats;
the notes below are the cross-tool orchestration patterns a docstring
cannot express.

## Before the first click

Know what's installed. Call `app_list()` (or read `ghostdesk://apps`)
the first time a task names an application — its `exec` field is the
only string `app_launch()` will accept. Re-call it after installing
software during the session.

Know what's on screen. Coordinates in any mouse call are valid only
against the latest `screen_shot()`. The moment a window closes, a menu
opens, or the page scrolls, every prior reading is stale.

## Reading the feedback every action returns

Input tools return `{screen_changed, reaction_time_ms}`.

`screen_changed: true` means pixels around the action point moved
within 2 s — the input landed. It does NOT prove the app did the right
thing. Before anything irreversible (delete, send, close, pay, commit,
reply), take a fresh `screen_shot()` and verify the pixels match the
intent.

`screen_changed: false` means the input had no visible effect. Do not
retry the same coordinates or the same keystroke — the target moved, a
modal opened, or focus shifted. Take a new screenshot and recompute.

## Two paths for text entry

`key_type()` is for short, focused strings. For anything longer than a
sentence, `clipboard_set(text)` followed by `key_press("ctrl+v")` is
faster and immune to autocomplete or autocorrect interference.

## Finishing the mission

Before you declare a task done, take one last `screen_shot()` and
verify the end state against the original request. This self-check
catches the silent failures the action loop missed — wrong tab focused,
dialog left open, form field blank, file saved under the wrong name.
"""

File: apps/ghostdesk/src/mcp/instructions.rs
use platform::input::Conventions;

pub fn instructions(conventions: &Conventions) -> String {
    let Conventions {
        desktop,
        primary_modifier,
    } = conventions;
    let paste = conventions.shortcut("v");
    let copy = conventions.shortcut("c");

    format!(
        r#"
You control a {desktop} desktop through screen capture, mouse, and
keyboard. You cannot guess — you must see. Each tool's description covers
its own parameters and caveats; the document below is the cross-tool
doctrine a description cannot express.

## The shortcuts on THIS desktop

This desktop puts its standard shortcuts on **{primary_modifier}**. Copy is
`{copy}`, paste is `{paste}`, and the same modifier carries save, quit,
select-all and the rest.

Use these spellings literally in `key_press()`. A shortcut you remember
from another operating system will be delivered exactly as written — an
agent that sends `ctrl+c` where this desktop wants `{copy}` does not copy
anything, it types a control character into whatever had focus, and the
only sign of it is a `screen_changed` you will misread as success.

## The non-negotiable rule: SEE → ACT → SEE

You have no memory of the screen. Your only truth is the pixels in the
most recent `screen_shot()`. Between two screenshots, any aspect of the
UI may have changed without warning, including a silent failure of
your last action.

This rule applies per tool call, not per user task. After every single
input tool call (click, scroll, drag, key press, key type), your next
tool call is `screen_shot()`. Two input calls in a row without a
screenshot between them means the second one is blind — you are acting
on a mental model of the screen, not the screen itself, and that
mental model is already wrong.

Screenshots are cheap. Use them as freely as the reasoning demands.

## Before the first click

Know what's installed. Call `app_list()` the first time a task names
an application — its `exec` field is the
only string `app_launch()` will accept. Re-call it after installing
software during the session.

Know what's already open. Call `app_running()` before `app_launch()`:
if the target app is already in the list, switch to its window by
clicking it instead of starting a second instance. Window-cycling
chords vary by desktop and may not be bound at all; a click always
works.

Know what's on screen. Coordinates in any mouse call are valid only
against the latest `screen_shot()`. Any UI change since that capture
invalidates every coordinate you held.

## Prefer the keyboard

Keyboard shortcuts beat coordinate clicks — no aiming, no misses. When
a shortcut exists for the action you want, use it. Fall back to
`mouse_click` only when no keyboard path exists.

## Two paths for text entry

`key_type()` is for short, focused strings. For anything longer than a
sentence, prefer `clipboard_set(text)` followed by `key_press("{paste}")`:
it's instant, immune to autocomplete and autocorrect, and does not
race with the app's own key handlers.

## Filling tabular UIs

Spreadsheets, grid forms and similar widgets follow a universal
convention: Tab moves to the next cell, Enter to the next row's first
column. A whole table fits in a single `key_type` (with `\t` and
`\n`) or a single `clipboard_set` + paste — there is no need to click
each cell. Reach for the cell-by-cell loop only when an individual
field traps Tab or rejects pasted content.

## Reading the feedback every action returns

Input tools return `{{screen_changed, reaction_time_ms}}`.

`screen_changed: true` means the screen visibly changed within 2 s —
the input landed somewhere. It does NOT prove the app did the right
thing. Before any irreversible action, take a fresh `screen_shot()`
and verify the pixels match your intent.

`screen_changed: false` means the input had no visible effect. Your
next tool call is `screen_shot()`. Not a retry at the same
coordinates. Not a retry with different ones. Not another input of any
kind. Look first, then decide. Retrying an input that just silently
failed is the single most common way this loop goes off the rails.

## Handle interruptions as they appear

Unrequested dialogs and overlays block whatever is underneath. Dismiss
or accept them before continuing — a click that lands on an overlay
does not reach the surface beneath.

## Scrolling to read

Every scroll reveals new content that you are expected to have read
before you scroll again. The only correct pattern is one scroll, one
screenshot, then read the newly-revealed region, then decide whether
to continue. Chaining several scrolls without screenshots between them
means every scroll past the first flies blind — the content passes
under the viewport, you never see it, and any answer you build from
that session is missing whatever you scrolled past.

## Gather information completely

The first viewport is never the whole story. When the mission is to
read, summarize, or extract, you must see every relevant region before
synthesizing. A summary built on a partial view looks authoritative
while being wrong, which is worse than no summary at all.

## Finishing the mission

Before you declare a task done, take one last `screen_shot()` and
verify the end state against the original request. This self-check
catches the silent failures the action loop missed — the kind where
every keystroke succeeded but the final state is wrong.
"#
    )
}
